cpp/lib/ClientConfigParser.cpp

```diff
@@ -0,0 +1,292 @@
+    {
+      CXX_LOG_INFO("Using client configuration path from binary directory: %s", binaryDirFilePath.c_str());
+      return binaryDirFilePath;
+    }
+    else
+    {
+      // 4. Try user home dir
+      return resolveHomeDirConfigPath();
+    }
+  }
+}
+
+// Private =====================================================================
+////////////////////////////////////////////////////////////////////////////////
+std::string EasyLoggingConfigParser::resolveHomeDirConfigPath()
+{
+  char envbuf[MAX_PATH + 1];
+#if defined(WIN32) || defined(_WIN64)
+  std::string homeDirFilePath;
+  char* homeDir;
+  if ((homeDir = sf_getenv_s("USERPROFILE", envbuf, sizeof(envbuf))) && strlen(homeDir) != 0)
+  {
+    homeDirFilePath = homeDir + PATH_SEP + SF_CLIENT_CONFIG_FILE_NAME;
+  } else {
+    // USERPROFILE is empty, try HOMEDRIVE and HOMEPATH
+    char* homeDriveEnv = sf_getenv_s("HOMEDRIVE", envbuf, sizeof(envbuf));
+    char* homePathEnv = sf_getenv_s("HOMEPATH", envbuf, sizeof(envbuf));
+    if (homeDriveEnv && strlen(homeDriveEnv) != 0 && homePathEnv && strlen(homePathEnv) != 0)
+    {
+      homeDirFilePath = std::string(homeDriveEnv) + homePathEnv + PATH_SEP + SF_CLIENT_CONFIG_FILE_NAME;
+    }
+  }
+  if (boost::filesystem::is_regular_file(homeDirFilePath))
+  {
+    CXX_LOG_INFO("Using client configuration path from home directory: %s", homeDirFilePath.c_str());
+    return homeDirFilePath;
+  }
+#else
+  char* homeDir;
+  if ((homeDir = sf_getenv_s("HOME", envbuf, sizeof(envbuf))) && (strlen(homeDir) != 0))
+  {
+    std::string homeDirFilePath = std::string(homeDir) + PATH_SEP + SF_CLIENT_CONFIG_FILE_NAME;
+    if (boost::filesystem::is_regular_file(homeDirFilePath))
+    {
+      CXX_LOG_INFO("Using client configuration path from home directory: %s", homeDirFilePath.c_str());
+      return homeDirFilePath;
+    }
+  }
+#endif
+  return "";
+}
+
+////////////////////////////////////////////////////////////////////////////////
+void EasyLoggingConfigParser::parseConfigFile(
+  const std::string& in_filePath,
+  client_config& out_clientConfig)
+{
+  value jsonConfig;
+  std::string err;
+  std::ifstream configFile;
+  try
+  {
+    configFile.open(in_filePath, std::fstream::in | std::ios::binary);
+    if (!configFile)
+    {
+      CXX_LOG_INFO("Could not open a file. The file may not exist: %s",
+        in_filePath.c_str());
+      std::string errMsg = "Error finding client configuration file: " + in_filePath;
+      throw ClientConfigException(errMsg.c_str());
+    }
+#if !defined(WIN32) && !defined(_WIN64)
+    checkIfValidPermissions(in_filePath);
+#endif
+    err = parse(jsonConfig, configFile);
+
+    if (!err.empty())
+    {
+      CXX_LOG_ERROR("Error in parsing JSON: %s, err: %s", in_filePath.c_str(), err.c_str());
+      std::string errMsg = "Error parsing client configuration file: " + in_filePath;
+      throw ClientConfigException(errMsg.c_str());
+    }
+  }
+  catch (std::exception& e)
```

Comment:
We don't need to explicitly close the file. It will close automatically when destructor of ifstream is called:
- https://en.cppreference.com/w/cpp/language/raii
- https://en.cppreference.com/w/cpp/io/basic_ifstream (destructor doc)